server/src/main/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesAction.java

```diff
@@ -0,0 +1,134 @@
+        super(settings, GetDiscoveredNodesAction.NAME, transportService, actionFilters,
+            (Reader<GetDiscoveredNodesRequest>) GetDiscoveredNodesRequest::new);
+
+        this.transportService = transportService;
+        if (discovery instanceof Coordinator) {
+            coordinator = (Coordinator) discovery;
+        } else {
+            coordinator = null;
+        }
+    }
+
+    @Override
+    protected void doExecute(Task task, GetDiscoveredNodesRequest request, ActionListener<GetDiscoveredNodesResponse> listener) {
+        if (coordinator == null) { // TODO remove when not nullable
+            throw new IllegalStateException("discovered nodes are not exposed by this discovery type");
+        }
+
+        final DiscoveryNode localNode = transportService.getLocalNode();
+        assert localNode != null;
+        if (localNode.isMasterNode() == false) {
+            throw new ElasticsearchException("this node is not master-eligible");
+        }
+        final AtomicBoolean listenerNotified = new AtomicBoolean();
+        final ListenableFuture<GetDiscoveredNodesResponse> listenableFuture = new ListenableFuture<>();
+        final ThreadPool threadPool = transportService.getThreadPool();
+        listenableFuture.addListener(listener,
+            threadPool.generic(),
+            threadPool.getThreadContext());
+
+        final Consumer<Iterable<DiscoveryNode>> respondIfRequestSatisfied = new Consumer<Iterable<DiscoveryNode>>() {
+            @Override
+            public void accept(Iterable<DiscoveryNode> nodes) {
+                final Set<DiscoveryNode> nodesSet = new LinkedHashSet<>();
+                nodesSet.add(localNode);
+                nodes.forEach(nodesSet::add);
+                logger.trace("discovered {}", nodesSet);
+                if (nodesSet.size() >= request.getWaitForNodes() && listenerNotified.compareAndSet(false, true)) {
+                    listenableFuture.onResponse(new GetDiscoveredNodesResponse(nodesSet));
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "waiting for " + request;
+            }
+        };
+
+        final Releasable releasable = coordinator.withDiscoveryListener(respondIfRequestSatisfied);
+        listenableFuture.addListener(finallyListener(releasable::close), threadPool.generic(), threadPool.getThreadContext());
+
+        threadPool.schedule(request.getTimeout(), Names.GENERIC, new Runnable() {
+            @Override
+            public void run() {
+                if (listenerNotified.compareAndSet(false, true)) {
+                    listenableFuture.onFailure(new ElasticsearchTimeoutException("timed out while waiting for " + request));
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "timeout handler for " + request;
+            }
+        });
+
+        respondIfRequestSatisfied.accept(coordinator.getFoundPeers());
```

Comment:
Yes, fixed this somewhere after a test failed. If `getFoundPeers` throws an exception (I don't think it can at the moment) then yes `TransportAction` sends this to the listener.